Added functionality for Gordo and Guillermo

blackshadowshade opened the pull request:
Fixes #265.

This pull request enables Gordo and Guillermo. It enforces the uniqueness of all die sizes for Gordo, and just swing values for Guillermo. It also stops players from accepting an aux die for Gordo that is a swing die with a type already in Gordo's recipe.

I've tested this with Gordo vs Guillermo, Gordo vs King Arthur, and Gordo vs Lancelot.

In addition, I've removed an unused argument in Giant's button special, which I noticed when I was using it as a template for these button specials.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1398/ (if it passes)

blackshadowshade commented:
Heads up for @Hurtay that Gordo should be available soon.  :)

cgolubi1 commented:
This didn't pass Jenkins.

blackshadowshade commented:
http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1400/ (if it passes)

cgolubi1 commented:
Okay, let me see if i can just use the replay test engine without modification to look for regressions here.  This seems like a straightforward application of replay testing.

cgolubi1 commented:
Oh, hmm, i seem to have gotten confused about this --- i tried replay testing, and it failed because it doesn't know how to deal with Gordo wanting unique swing values.  But in fact what i should be doing is temporarily skipping Gordo and Guillermo, and seeing if the rest of the replay testing works.

Then before merging i'll need to make the client support those buttons.

cgolubi1 commented:
Okay, i replayed about 4000 games without Gordo and Guillermo, and saw no problems.  Next step is to make RandomAI able to support buttons with the unique skill.

blackshadowshade commented:
Would you be able to put up a dev site for this so that the testers can get onto it?

cgolubi1 commented:
Sure, will do.

cgolubi1 commented:
irilyth: can you sync up 'temporary' to 54.167.234.232 in route53?

irilyth commented:
> irilyth: can you sync up 'temporary' to 54.167.234.232 in route53?

Done!

dwvanstone commented:
This is great! I've done some testing up on 54.167.234.232.

This is the only issue I've found:

 Gordo v. Michigan, accepting auxiliary dice. When selecting swing dice, Gordo has (V) (W) (X) (Y) (Z) (6). If I select V=6, W=7, X=8, Y=9, Z=10, it gives me an error: "The swing values chosen are not unique." Perhaps a clearer error would be "The swing values must be unique and not be the same size as any other dice." (I.e., it should be a different error from when Guillermo has non-unique swing dice.)

blackshadowshade commented:
After a great deal of consideration of how to reword the message, with the proviso that it not be very long (and thus take up vertical real estate on small screens), I've changed the message to 

"Cannot have multiple dice with the same size"

Are you okay with that, @dwvanstone?

@cgolubi1, I've got a Jenkins run for this:

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1438/ (if it passes)

However, I've neither rebased nor merged master. Once we're done with testing, I'll rebase this and run a fresh Jenkins run.

cgolubi1 commented:
Sounds good - thanks for not rebasing - i'll update the dev site.

cgolubi1 commented on src/engine/BMGame.php:
```diff
@@ -912,7 +912,11 @@ protected function set_swing_values() {
                 // check for unique swing values
                 if (!self::are_unique_swing_values_valid($player)) {
-                    $this->message = 'The swing values chosen are not unique';
+                    if ('Gordo' == $player->button->name) {
```
Hmm, wait... i'm not going to update the dev site for now, because i'm confused about how this fix is relevant, and i think we can talk about it on this issue just as easily.  Does this have anything to do with whether the button is Gordo?  I thought the problem was that the opposing button had an auxiliary die.  Why would Guillermo not have the same problem?

cgolubi1 replied:
Oh, i see what you're saying from the original statement of the pull --- the rules are actually different for Gordo vs Guillermo.  Well, okay, i will update the dev site then.  Is there really no better way to identify buttons that work like Gordo, though?  Grepping by button name deep in the backend seems really clunky.

blackshadowshade replied:
Last night, I was thinking that it would be okay, it's only a once-off. However, thinking about it now, I think I agree that it's a bit clunky. I'll see if I can implement some sort of hooked function to move the logic to the BtnSkill classes.

dwvanstone commented:
> "Cannot have multiple dice with the same size"

Yes, that is perfect. Thanks!

blackshadowshade commented:
Right, that logic has been shifted into a button hook.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1439/ (if it passes)

cgolubi1 commented:
Thanks.  I'll update the dev site (should be a noop) and the replay site i'm working on.

cgolubi1 commented:
Worth noting is something i realised when i was making the change for #2305 --- unless i'm missing something, there's no API way to find out what button skills a particular button has (you can tell which button skills are in the game, but the API doesn't report which button has the skills).  So API clients/bot players are going to have to do what RandomAI does here, and look at the player's button name.  I suppose i could also check if Gordo or Guillermo is in the button skills just for fun, but that seems like extra work for no benefit, since the button skills are named after the buttons.

cgolubi1 commented:
Anyway.  I'm going to okay this for merging once my replay site playes a few hundred Gordo or Guillermo games without falling over, which is what it's trying right now.  I'm taking dwvanstone's comments above as approval, so i think we're good to go here.

cgolubi1 commented:
```
Starting game 24
Traceback (most recent call last):
  File "./test_log_games", line 71, in <module>
    c.log_test_game(i, button1, button2)
  File "lib/random_ai.py", line 1545, in log_test_game
    self.next_game_action()
  File "lib/random_ai.py", line 1533, in next_game_action
    elif state == 'CHOOSE_AUXILIARY_DICE': self._game_action_choose_auxiliary_dice()
  File "lib/random_ai.py", line 1406, in _game_action_choose_auxiliary_dice
    return self._game_action_choose_auxiliary_dice_player(chosen_client, chosen_array)
  File "lib/random_ai.py", line 1105, in _game_action_choose_auxiliary_dice_player  
    retval and retval.message or "NULL"))
  File "lib/random_ai.py", line 1444, in bug
    raise ValueError, message
ValueError: API choose_auxiliary_dice(24, add, 5) unexpectedly failed: This button cannot add this auxiliary die
```

The game is Lancelot vs Gordo.  How do we feel about that?  I feel that the behavior is okay --- i tried this game in the UI (see screencap) and it doesn't look that unclean.  But (1) obviously i'll have to make another change to RandomAI to work with this behavior, and (2) on the PR side, i think Gordo's flavor text should make it clear that this is something that can happen, so that when players see this error, they can mouseover the button special description and get a hint about what happened.

cgolubi1 commented:
Okay, i have a random_ai.py modification in a working directory that has played 100 games of Gordo vs. Auxiliary without blowing up, and wound up choosing the Aux in a small handful of them (spot-check: Gordo vs. Jota with the +(12), which sounds legitimate).  I'll go back to full replay testing with that change, and assuming it looks good, i'll commit it.

I'm still requesting the modification to the button skill description clarifying that the Gordo player is required to decline an auxiliary die which would have the same size as another die.

blackshadowshade commented:
I've updated the description. I've also rebased this because I'm about to head off on holidays, and I didn't want to hold up accepting this pull request.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1442/ (if it passes)

cgolubi1 commented:
LGTM.  I'll check in on my replay site.

cgolubi1 commented:
It's doing great; i'll submit a PR for my random_ai change.

cgolubi1 commented:
Passes, and the only change is the mentioned description one:
```
79c79,81
<         return 'None of Gordo\'s dice can be the same size.';
---
>         return 'None of Gordo\'s dice can be the same size. ' .
>                'Note that a player playing Gordo is required to decline an auxiliary die ' .
>                'that would force two of Gordo\'s dice to be the same size';
```
